crates/polars-sql/src/function_registry.rs

```diff
@@ -0,0 +1,87 @@
+//! This module defines the function registry and user defined functions.
+use polars_arrow::error::{polars_bail, PolarsResult};
+use polars_core::prelude::Field;
+use polars_core::schema::Schema;
+use polars_core::series::Series;
+pub use polars_plan::prelude::{Context, FunctionOptions};
+use polars_plan::prelude::{Expr, FunctionOutputField, GetOutput, SeriesUdf};
+/// A registry that holds user defined functions.
+pub trait FunctionRegistry: Send + Sync {
+    /// Register a function.
+    fn register(&mut self, name: &str, fun: &dyn UserDefinedFunction) -> PolarsResult<()>;
+    /// Call a user defined function.
+    fn call(&self, name: &str, args: Vec<Expr>) -> PolarsResult<Expr>;
+    /// Check if a function is registered.
+    fn contains(&self, name: &str) -> bool;
+}
+
+/// A default registry that does not support registering or calling functions.
+pub struct DefaultFunctionRegistry {}
+
+impl FunctionRegistry for DefaultFunctionRegistry {
+    fn register(&mut self, _name: &str, _fun: &dyn UserDefinedFunction) -> PolarsResult<()> {
+        polars_bail!(ComputeError: "'register' not implemented on DefaultFunctionRegistry'")
+    }
+
+    fn call(&self, _name: &str, _args: Vec<Expr>) -> PolarsResult<Expr> {
+        polars_bail!(ComputeError: "'call_udf' not implemented on DefaultFunctionRegistry'")
+    }
+    fn contains(&self, _name: &str) -> bool {
+        false
+    }
+}
+
+/// Represents a user-defined function that can be used within the system.
+///
+/// This trait provides several methods to describe and manipulate the user-defined function.
+/// This is to be used with the [`FunctionRegistry`].
+pub trait UserDefinedFunction: Send + Sync {
```

Comment:
Now we have 2 traits that define a UDF. Can we rewrite this to use `SeriesUdf`. We should consolidate functionality.